File: Dynamic_Programming/number-of-dice-roles-with-target-sum.cpp
class Solution {
public:
    
    long long mod = 1e9+7;
    
    int solve(int n,int k,int t,vector<vector<long long>> &dp)
    {
        if(t<0)
            return 0;
        if(n == 0 && t != 0)
            return 0;
        if(n!=0 && t == 0)
            return 0;
        if(n == 0 && t == 0)
            return 1;
        
        if(dp[n][t] != -1)
            return dp[n][t];
        
        long long ans = 0;
        
        for(int i=1;i<=k;i++)
        {
            ans = (ans%mod + solve(n-1,k,t-i, dp)%mod)%mod;
        }
        return dp[n][t] = ans;
        
    }
    
    
    
    
    int numRollsToTarget(int n, int k, int target) {
     
         vector<vector<long long>> dp(n+1, vector<long long>(target+1, -1));
        return solve(n, k, target, dp);
        
    }
};

File: Dynamic_Programming/number-of-dice-roles-with-target-sum.py
mod = 10**9+7
class Solution:
    def numRollsToTarget(self, n: int, k: int, target: int) -> int:
        dp = [[0 for i in range(target+1)] for j in range(n+1)]
        
        dp[0][0] = 1
        
        for i in range(1,n+1):
            for j in range(i,min(k*i,target)+1):
                if i == 1:
                    dp[i][j] = 1
                else:
                    en = j-1
                    st = max(1,j-k)
                    dp[i][j] = sum(dp[i-1][st:en+1])
                    
        #print(dp)
        return dp[n][target]%mod
                    
